src/middlewared/middlewared/plugins/cloud_sync.py

```diff
@@ -0,0 +1,522 @@
+        return await self.middleware.call('datastore.query', 'system.cloudcredentials', filters, options)
+
+    @accepts(Dict(
+        'cloud_sync_credentials',
+        Str('name'),
+        Str('provider'),
+        Dict('attributes', additional_attrs=True),
+        register=True,
+    ))
+    async def do_create(self, data):
+        self._validate("cloud_sync_credentials", data)
+
+        data["id"] = await self.middleware.call(
+            'datastore.insert',
+            'system.cloudcredentials',
+            data,
+        )
+        return data
+
+    @accepts(Int('id'), Ref('cloud_sync_credentials'))
+    async def do_update(self, id, data):
+        self._validate("cloud_sync_credentials", data)
+
+        await self.middleware.call(
+            'datastore.update',
+            'system.cloudcredentials',
+            id,
+            data,
+        )
+
+        return data
+
+    @accepts(Int('id'))
+    async def do_delete(self, id):
+        await self.middleware.call(
+            'datastore.delete',
+            'system.cloudcredentials',
+            id,
+        )
+
+    def _validate(self, schema_name, data):
+        provider = REMOTES[data["provider"]]
+
+        verrors = ValidationErrors()
+
+        if data["provider"] not in REMOTES:
+            verrors.add(f"{schema_name}.provider", "Invalid provider")
+        else:
+            attributes_verrors = validate_attributes(provider.credentials_schema, data)
+            verrors.add_child(f"{schema_name}.attributes", attributes_verrors)
+
+        if verrors:
+            raise verrors
+
+
+class CloudSyncService(CRUDService):
+
+    class Config:
+        datastore = 'tasks.cloudsync'
+        datastore_extend = 'cloudsync._extend'
+
+    @private
+    async def _extend(self, cloud_sync):
+        cloud_sync['credentials'] = cloud_sync.pop('credential')
+
+        if 'encryption_password' in cloud_sync:
+            cloud_sync['encryption_password'] = await self.middleware.call(
+                'notifier.pwenc_decrypt', cloud_sync['encryption_password'])
+        if 'encryption_salt' in cloud_sync:
+            cloud_sync['encryption_salt'] = await self.middleware.call(
+                'notifier.pwenc_decrypt', cloud_sync['encryption_salt'])
+
+        return cloud_sync
+
+    @private
+    async def _compress(self, cloud_sync):
+        if 'credentials' in cloud_sync:
+            cloud_sync['credential'] = cloud_sync.pop('credentials')
+
+        if 'encryption_password' in cloud_sync:
+            cloud_sync['encryption_password'] = await self.middleware.call(
+                'notifier.pwenc_encrypt', cloud_sync['encryption_password'])
+        if 'encryption_salt' in cloud_sync:
+            cloud_sync['encryption_salt'] = await self.middleware.call(
+                'notifier.pwenc_encrypt', cloud_sync['encryption_salt'])
+
+        return cloud_sync
+
+    @private
+    async def _get_cloud_sync(self, id):
```

Comment:
We already have _get_instance in CRUDService that does exactly that, if I am not mistaken?